### PR TITLE
Add support for Roslyn static analysis - FxCop like rules

### PR DESCRIPTION
**Pull request by bgavrilMS**

### CODE REVIEW

**Comment by henrik-me** [comment]
Does the new ruleset have all the SDL rules enabled? Also the configureawait stuff?

**Comment by henrik-me** [comment]
I assume given it's solution wide setting that it works on all projects (including the test ones) ?

**Review by henrik-me — CHANGES_REQUESTED** [review]
:clock1:

**Comment by bgavrilMS** [comment]
I have kept the configure await rules. I haven't looked at enabling any SDL rules, not sure there are Roslyn analyzers for those. Product code uses the rules, tests and samples do not (this is like the status quo, the unit test projects were using a ruleset that had all the async rules disabled). Would you find adding the same rules to the test and samples? I think we can have a different ruleset, especially on the samples.

---
In reply to: [429914428](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1325#issuecomment-429914428) [](ancestors = 429914428)

**Comment by bgavrilMS** [comment]
I have configured that ruleset to be used our 3 product projects.

---
In reply to: [429914687](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1325#issuecomment-429914687) [](ancestors = 429914687)

File: msal/src/Common-Old/AuthenticationResult.cs [hunk 1]
```diff
@@ -1,140 +0,0 @@
...
-
-using System;
-using System.Runtime.Serialization;
-
-namespace Microsoft.IdentityModel.Clients.ActiveDirectory
-{
-    /// <summary>
-    /// Represents the outcome of one authentication operation.
-    /// </summary>
-    public enum AuthenticationStatus
```
Comment (bgavrilMS) [inline msal/src/Common-Old/AuthenticationResult.cs:36]: >public [](start = 4, length = 6)

Old code, I doubt it is useful anymore. #Resolved

File: core/ConfigureAwaitDisable.ruleset [hunk 2]
```diff
@@ -1,77 +0,0 @@
...
-    <Description Resource="MinimumRecommendedRules_Description" />
-  </Localization>
-  <Rules AnalyzerId="AsyncUsageAnalyzers" RuleNamespace="AsyncUsageAnalyzers">
-    <Rule Id="AvoidAsyncSuffix" Action="None" />
-    <Rule Id="AvoidAsyncVoid" Action="None" />
-    <Rule Id="UseAsyncSuffix" Action="None" />
-    <Rule Id="UseConfigureAwait" Action="None" />
-  </Rules>
-  <Rules AnalyzerId="Microsoft.Analyzers.ManagedCodeAnalysis" RuleNamespace="Microsoft.Rules.Managed">
-    <Rule Id="CA1001" Action="Warning" />
```
Comment (bgavrilMS) [inline core/ConfigureAwaitDisable.ruleset:14]: >CA1001 [](start = 14, length = 6)

I did not manage to get the built in rules to work, so I doubt these were ever doing anything for us...

File: msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 3]
```diff
@@ -182,29 +183,9 @@
     <None Update="Platforms\net45\WindowsFormsWebAuthenticationDialogBase.cs" SubType="Form" />
     <None Update="Platforms\net45\SilentWindowsFormsAuthenticationDialog.cs" SubType="Form" />
   </ItemGroup>
-
-  <!-- Copies all artifacts needed to generate the nuget package. This should be replaced with msbuild /t:pack, however
```
Comment (bgavrilMS) [inline msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:186]: Not needed, we package via /t:pack nowadays #Resolved

File: msal/tests/dev apps/XForms/XForms/XForms.csproj [hunk 4]
```diff
@@ -3,8 +3,7 @@
     <TargetFramework>netstandard1.3</TargetFramework>
     <PackageTargetFallback>portable-net45+win8+wpa81+wp8</PackageTargetFallback>
     <TreatWarningsAsErrors>false</TreatWarningsAsErrors>
-    <RunCodeAnalysis>false</RunCodeAnalysis>
-    <CodeAnalysisRuleSet>..\..\..\..\build\Sdl7.0.ruleset</CodeAnalysisRuleSet>
```
Comment (bgavrilMS) [inline msal/tests/dev apps/XForms/XForms/XForms.csproj:7]: >Sdl7 [](start = 43, length = 4)

This was the only project that was using this SDL ruleset.

File: msal/build/Sdl7.0.ruleset [hunk 5]
```diff
@@ -1,143 +0,0 @@
-﻿<?xml version="1.0" encoding="utf-8"?>
-<RuleSet Name="SDL 7.0" Description="SDL 7.0-prescribed required and recommended checks." ToolsVersion="11.0">
-  <Rules AnalyzerId="Microsoft.Analyzers.ManagedCodeAnalysis" RuleNamespace="Microsoft.Rules.Managed">
```
Comment (bgavrilMS) [inline msal/build/Sdl7.0.ruleset:3]: >ManagedCodeAnalysis [](start = 41, length = 19)

This ruleset was used only by a test project. Note that SDL rules do run as part of the build, but through the old FxCop (i.e. binary level analysis, which, arguably, is more powerful)

**Comment by henrik-me** [comment]
IMO, same standards apply too all of our code. Including samples: we are also responsible for security/accessibility etc. for samples

---
In reply to: [429920981](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1325#issuecomment-429920981) [](ancestors = 429920981,429914428)

**Comment by henrik-me** [comment]
I would have liked it everywhere

---
In reply to: [429921139](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1325#issuecomment-429921139) [](ancestors = 429921139,429914687)

File: msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj [hunk 6]
```diff
@@ -182,29 +183,9 @@
...
-    <Copy SourceFiles="@(FilesToPack)" DestinationFiles="$(NugetPackageRootFolder)\lib\%(RecursiveDir)%(FileName)%(Extension)" />
-    <Copy SourceFiles="$(NuspecFilePath)" DestinationFolder="$(NugetPackageRootFolder)" />
-  </Target>
-
-  <PropertyGroup>
-    <CodeAnalysisRuleSet>../../../core/ConfigureAwait.ruleset</CodeAnalysisRuleSet>
-  </PropertyGroup>
+   
   <ItemGroup>
     <PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" PrivateAssets="All" />
```
Comment (henrik-me) [inline msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:188]: ><PackageReference Include="AsyncUsageAnalyzers" Version="1.0.0-alpha003" PrivateAssets="All" /> [](start = 4, length = 95)

do we then still need this one? It's not part of the Roslyn analyzers?

**Review by henrik-me — APPROVED** [review]
:shipit:

File: msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj (hunk 6 above)
Comment (bgavrilMS) [reply msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:188]: We use 2 rules from AsyncUsageAnalyzers - 1. ensure ConfigureAwait is explicitly set and 2. Use Async suffix. 

The new analyzers have support for //2 but not for //1 so I decided to keep the AsyncUsageAnalyzers

**Comment by bgavrilMS** [comment]
This PR is now complete but will not merge the work into dev until release.

**Comment by henrik-me** [comment]
great

---
In reply to: [430195790](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1325#issuecomment-430195790) [](ancestors = 430195790)

**Pull request merged**